### PR TITLE
fix: centralize YAML I/O with UTF-8 encoding for cross-platform safety

### PR DESCRIPTION
**Pull request by danielmeppiel**
## Description

Adds a centralized `yaml_io.py` module (`load_yaml`, `dump_yaml`, `yaml_to_str`) that guarantees UTF-8 encoding and `allow_unicode=True` on all YAML file operations. This prevents silent mojibake on Windows where the default file encoding is cp1252.

**Root cause**: PyYAML's default `allow_unicode=False` escapes non-ASCII characters as `\xNN` sequences, and many `open()` calls lacked explicit `encoding='utf-8'`. On Windows (cp1252 default encoding), this created a write-read mismatch that would silently corrupt non-ASCII data like accented author names.

Addresses the root cause behind #387.

## Changes

- **New module**: `src/apm_cli/utils/yaml_io.py` -- 3 functions, ~55 LOC
- **Migrated 20 YAML I/O sites** across 15 source files to use the new helpers
- **CI guard**: grep-based step in `ci.yml` catches any `yaml.dump/safe_dump` writing to file handles outside `yaml_io.py`
- **18 new tests** in `tests/unit/test_yaml_io.py` covering UTF-8 round-trips, unicode preservation, cross-platform safety, key ordering, edge cases

## Type of change

- [x] Bug fix
- [x] Maintenance / refactor

## Testing

- [x] All 2,676 unit tests pass
- [x] 18 new yaml_io-specific tests pass
- [x] CI encoding guard passes locally (zero violations)

## Relationship to #388

PR #388 by @alopezsanchez correctly identified the `allow_unicode` bug but only fixed write sites -- leaving read sites vulnerable to mojibake on Windows. This PR fixes the complete I/O path structurally. Once merged, #388 can be closed or rebased as a much smaller diff.

### CODE REVIEW

File: src/apm_cli/commands/uninstall/engine.py [hunk 1]
```diff
@@ -179,8 +179,8 @@ def _cleanup_transitive_orphans(lockfile, packages_to_remove, apm_modules_dir, a
     # Determine remaining deps to avoid removing still-needed packages
     remaining_deps = builtins.set()
     try:
-        with open(apm_yml_path, "r") as f:
-            updated_data = yaml.safe_load(f) or {}
+        from ...utils.yaml_io import load_yaml
+        updated_data = load_yaml(apm_yml_path) or {}
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/commands/uninstall/engine.py:182]: Fixed in 3b4b9e3 -- removed unused `import yaml`.

File: src/apm_cli/integration/mcp_integrator.py [hunk 2]
```diff
@@ -895,8 +895,8 @@ def install(
 
                     apm_yml = Path("apm.yml")
                     if apm_yml.exists():
-                        with open(apm_yml, "r", encoding="utf-8") as f:
-                            apm_config = yaml.safe_load(f)
+                        from apm_cli.utils.yaml_io import load_yaml
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/integration/mcp_integrator.py]: Fixed in 3b4b9e3 -- removed unused `import yaml` from the try-block.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -39,6 +39,21 @@ jobs:
     - name: Install dependencies
       run: uv sync --extra dev --extra build
 
+    - name: Check YAML encoding safety
+      run: |
+        # Ensure YAML file I/O goes through yaml_io helpers.
+        # Catches yaml.dump/safe_dump writing to a file handle outside yaml_io.py.
+        VIOLATIONS=$(grep -rn --include='*.py' \
+          -E 'yaml\.(safe_)?dump\(.+,\s*[a-zA-Z_]+\b' src/apm_cli/ \
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply .github/workflows/ci.yml]: Valid. Fixed in 3b4b9e3 -- switched from `grep -E` to `grep -P` (PCRE) which supports `\s` and `\b` natively. Ubuntu CI runners have GNU grep with PCRE support.

File: src/apm_cli/utils/yaml_io.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def load_yaml(path: Union[str, Path]) -> Optional[Dict[str, Any]]:
+    """Load a YAML file with explicit UTF-8 encoding.
+
+    Returns parsed data or ``None`` for empty files.
+    Raises ``FileNotFoundError`` or ``yaml.YAMLError`` on failure.
+    """
+    with open(path, "r", encoding="utf-8") as fh:
+        return yaml.safe_load(fh)
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/utils/yaml_io.py:35]: Intentional. Every `apm.yml` file is a YAML mapping (dict). Typing the return as `Dict | None` rather than `Any` gives callers type safety without needing a cast. If we ever load non-mapping YAML (e.g. a list), we'd use `yaml.safe_load()` directly since it wouldn't be an apm config file.

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Cross-platform YAML encoding: all `apm.yml` read/write operations now use explicit UTF-8 encoding via centralized `yaml_io` helpers, preventing silent mojibake on Windows cp1252 terminals; non-ASCII characters (e.g. accented author names) are preserved as real UTF-8 instead of `\xNN` escape sequences (#387, based on #388 by @alopezsanchez)
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply CHANGELOG.md]: Fair point. Fixed in 3b4b9e3 -- entry now ends with `(#387, #433)` and credits the contributor.

File: src/apm_cli/commands/_helpers.py [hunk 6]
```diff
@@ -457,5 +456,5 @@ def _create_minimal_apm_yml(config, plugin=False):
     apm_yml_data["scripts"] = {}
 
     # Write apm.yml
-    with open(APM_YML_FILENAME, "w") as f:
-        yaml.safe_dump(apm_yml_data, f, default_flow_style=False, sort_keys=False)
+    from ..utils.yaml_io import dump_yaml
+    dump_yaml(apm_yml_data, APM_YML_FILENAME)
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/commands/_helpers.py:458]: Fixed in 3b4b9e3 -- removed the unused `yaml = _lazy_yaml()` call.

File: src/apm_cli/commands/install.py [hunk 7]
```diff
@@ -82,8 +82,8 @@ def _validate_and_add_packages_to_apm_yml(packages, dry_run=False, dev=False, lo
 
     # Read current apm.yml
     try:
-        with open(apm_yml_path, "r") as f:
-            data = yaml.safe_load(f) or {}
+        from ..utils.yaml_io import load_yaml
+        data = load_yaml(apm_yml_path) or {}
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/commands/install.py:84]: Fixed in 3b4b9e3 -- removed unused `import yaml`.

File: src/apm_cli/compilation/agents_compiler.py [hunk 8]
```diff
@@ -73,8 +73,8 @@ def from_apm_yml(cls, **overrides) -> 'CompilationConfig':
             import yaml
             
             if Path('apm.yml').exists():
-                with open('apm.yml', 'r') as f:
-                    apm_config = yaml.safe_load(f) or {}
+                from ..utils.yaml_io import load_yaml
```
[the remark this answers is not in the document]
Comment (danielmeppiel) [reply src/apm_cli/compilation/agents_compiler.py:75]: Fixed in 3b4b9e3 -- removed both `import yaml` and the redundant inner `from pathlib import Path` (already imported at module level).
